Builder error when arg mutability is not well set in the receiver gen function

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 22.04 LTS
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CC version: cc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Git vroot status: weekly.2022.30-29-g0133ff2c
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /var/www/vproject/lew.test/lewlew/redis/Thai_foodchain_bug.v`
```v
module main

pub struct Redis {}

pub struct Reptile {
	name string
}

fn main() {
	mut r := Redis{}
	mut g := Reptile{
		name: 'Gecko'
	}
	r.node_create(mut g)
}

pub fn (mut r Redis) node_create<T>(t T) bool {
	return true
}

```

**What did you expect to see?**

A formatted error explain the mutability is not well set in the receiver function


**What did you see instead?**
```
==================
/tmp/v_1000/Thai_foodchain_bug.6649467286470948528.tmp.c:12153: error: cannot convert 'struct main__Reptile *' to 'struct main__Reptile'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1382,30 +1382,6 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 				final_arg_typ = exp_arg_sym.info.elem_type
 				final_arg_sym = c.table.sym(final_arg_typ)
 			}
-			if exp_arg_typ.has_flag(.generic) {
-				method_concrete_types := if method.generic_names.len == rec_concrete_types.len {
-					rec_concrete_types
-				} else {
-					concrete_types
-				}
-				if exp_utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
-					method_concrete_types)
-				{
-					exp_arg_typ = exp_utyp
-				} else {
-					continue
-				}
-
-				if got_arg_typ.has_flag(.generic) {
-					if got_utyp := c.table.resolve_generic_to_concrete(got_arg_typ, method.generic_names,
-						method_concrete_types)
-					{
-						got_arg_typ = got_utyp
-					} else {
-						continue
-					}
-				}
-			}
 			param := if method.is_variadic && i >= method.params.len - 1 {
 				method.params.last()
 			} else {
@@ -1440,6 +1416,30 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 						arg.expr.pos())
 				} else {
 					c.fail_if_unreadable(arg.expr, got_arg_typ, 'argument')
+				}
+			}
+			if exp_arg_typ.has_flag(.generic) {
+				method_concrete_types := if method.generic_names.len == rec_concrete_types.len {
+					rec_concrete_types
+				} else {
+					concrete_types
+				}
+				if exp_utyp := c.table.resolve_generic_to_concrete(exp_arg_typ, method.generic_names,
+					method_concrete_types)
+				{
+					exp_arg_typ = exp_utyp
+				} else {
+					continue
+				}
+
+				if got_arg_typ.has_flag(.generic) {
+					if got_utyp := c.table.resolve_generic_to_concrete(got_arg_typ, method.generic_names,
+						method_concrete_types)
+					{
+						got_arg_typ = got_utyp
+					} else {
+						continue
+					}
 				}
 			}
 			if left_sym.info is ast.Array && method_name == 'sort_with_compare' {

--- a/vlib/v/checker/tests/generics_method_arg_type_err.out
+++ b/vlib/v/checker/tests/generics_method_arg_type_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_method_arg_type_err.vv:14:20: error: `node_create` parameter `t` is not `mut`, `mut` is not needed`
+   12 |         name: 'Gecko'
+   13 |     }
+   14 |     r.node_create(mut g)
+      |                       ^
+   15 | }
+   16 |

--- a/vlib/v/checker/tests/generics_method_arg_type_err.vv
+++ b/vlib/v/checker/tests/generics_method_arg_type_err.vv
@@ -1,0 +1,19 @@
+module main
+
+pub struct Redis {}
+
+pub struct Reptile {
+	name string
+}
+
+fn main() {
+	mut r := Redis{}
+	mut g := Reptile{
+		name: 'Gecko'
+	}
+	r.node_create(mut g)
+}
+
+pub fn (mut r Redis) node_create<T>(t T) bool {
+	return true
+}

